Add overrides to allow multiple specs

kontura opened the pull request.

kontura committed:
Schemas are always strict in marshmallow 3

https://marshmallow.readthedocs.io/en/stable/upgrading.html#schemas-are-always-strict

kontura committed:
Remove unused spec_release decoding

kontura committed:
Allow spec_path and add patches_dir to DistGit

Some DistGits such as: https://git.centos.org/rpms/libsolv/tree/c8s have
spec in SPECS dir (we need spec_path) and patches in SOURCES dir (we
need patches_dir_path).

kontura committed:
Add override schema and class

It derives from the DistGit class and it allows specifying roots. The
override configuration is then used only for its roots.

Each component can have any number of overrides.

To be on the safe side it also changes `Component` constructor `requires`
argument to avoid mutable default.

kontura committed:
Add prep step and output for srpms from overrides

pep8speaks commented:
Hello @kontura! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:











There are currently no PEP 8 issues detected in this Pull Request. Cheers! :beers: 

##### Comment last updated at 2021-06-23 14:56:22 UTC

lukash commented on rgo/component.py:
```diff
@@ -30,7 +30,7 @@
 
 class Component(object):
-    def __init__(self, name, version_from=None, git=None, distgit=None, requires=[]):
+    def __init__(self, name, version_from=None, git=None, distgit=None, requires=[], overrides=[]):
```
The `overrides` override specifically the `distgit` dictionary, so I think they should either go under `distgit`, e.g.:
```
components:
  - name: foo
    git:
      src: github:someone/foo.git
    distgit:
      src: fedorapkgs:foo.git
      branch: origin/f32
      patches: drop
      overrides:
        - roots:
            - centos-stream-8-x86_64
          src: centos-distgit:foo.git
          branch: c8s
          spec-path: SPECS/foo.spec
```

If that's possible with the way you inherit the `Overrides` from `DistGit`.

Alternatively, call it `distgit-overrides` if we'd keep it under `component` in the yaml?

m-blaha replied:
Using mutable default arguments (`overrides=[]`) can have nasty side effects. We should probably avoid it. See https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/ for details.

lukash replied:
Oh, this. Yeah, if you don't mutate the value of the argument, its fine :) In this case it's "only" copied to a `set`.

But it might be better to avoid this altogether...

kontura replied:
> If that's possible with the way you inherit the `Overrides` from `DistGit`.
> 
> Alternatively, call it `distgit-overrides` if we'd keep it under `component` in the yaml?

It looks like its not possible so it will have to be `distgit-overrides`.

Just throwing an idea but I was thinking another option could be dropping the overrides concept and instead doing:
```
components:
  - name: foo
    git:
      src: github:someone/foo.git
    distgits:
      - src: fedorapkgs:foo.git
        branch: origin/f32
        patches: drop
      - src: centos-distgit:foo.git
        branch: c8s
        spec-path: SPECS/foo.spec
        roots:
          - centos-stream-8-x86_64
```
It is not backwards compatible (if that matters?) and I am not sure if no `roots` should mean all default builds.
Though it is simpler I think.

lukash replied:
I think `distgit-overrides` is more explicit, your suggestion would need the rule that if there are no `roots` it's the default, and then what if there are more items in the array that have no `roots`...

lukash commented on rgo/component.py:
```diff
@@ -95,94 +93,31 @@ def make_srpm(self, tmpdir):
-        spec_name = rpmspec.sourceHeader["Name"]
-        if isinstance(spec_name, bytes):
-            spec_name = spec_name.decode("utf-8")
+        original_spec_path = utils.extract_specfile_from_branch(workdir, spec_git, spec_name)
```
I see it as a bit of an anti-pattern to move code out to functions in `utils.py`, the functions are still specific to this module and not usable elsewhere, it would be much better to have them as private methods of this class?

lukash commented on rgo/component.py:
```diff
@@ -95,94 +93,31 @@ def make_srpm(self, tmpdir):
-            if src_type == rpm.RPMBUILD_ISPATCH:
-                if patches == PatchesAction.keep:
-                    _sources.append(src)
-            elif src_type == rpm.RPMBUILD_ISSOURCE:
```
This commit moves around a lot of code, I haven't really gone over all of it to check if it still does the same, but I noticed you've removed this part completely. IIUC this handles secondary sources listed in the spec file which is a marginal case, the main sources are supposed to be in the git and this is rarely used, correct?

It's good to point these things out in the commit message with a brief justification.

Any other "simplifications" you've done that I've missed? :slightly_smiling_face:

kontura replied:
> This commit moves around a lot of code, I haven't really gone over all of it to check if it still does the same, but I noticed you've removed this part completely. IIUC this handles secondary sources listed in the spec file which is a marginal case, the main sources are supposed to be in the git and this is rarely used, correct?

ah, I don't know, I didn't know secondary sources were a thing, I though its some not-working version of using tarball from distgit. Although even the secondary sources wouldn't work unless they were present in the distgit itself? or I have missed where they are downloaded.

> It's good to point these things out in the commit message with a brief justification.
> 
> Any other "simplifications" you've done that I've missed? slightly_smiling_face

My bad. I can think of only the way I changed keeping track of the build status, storing the `build_proxy` instead of the `state` and `success`, looking at it now that could be also a separate commit.

lukash replied:
> ah, I don't know, I didn't know secondary sources were a thing, I though its some not-working version of using tarball from distgit. Although even the secondary sources wouldn't work unless they were present in the distgit itself? or I have missed where they are downloaded.

I'm not sure, I'm fine dropping it, as we'd need to go out of our way to test it and yeah it probably wasn't working in the first place?

kontura replied:
ok, ok.
I have split the commit into two, it should be more manageable now.

lukash commented on rgo/component.py:
```diff
@@ -95,94 +93,31 @@ def make_srpm(self, tmpdir):
+                                                             self.version_from, override_git.patches)
+            if override_git.patches == PatchesAction.keep:
+                utils.prepare_specifle_patches(workdir, final_spec_path, override_git)
+            srpm_path = utils.rpmbuild_srpm(workdir, final_spec_path)
```
I like how you moved the code from here to the various functions, but ultimately there's still this bit that's repeated inside the `for` loop and also called on the main srpm before it. So this seems like this should go into a function (a private method of this class, preferably?), and then it's possibly not even necessary to maintain the other functions you've created, since they'll be only called from a single place. The full business logic could just be in that single method (it'll be a bit busy, but I personally prefer that to another level of indirection).

kontura replied:
I have started with that but then changed it to this, I like how the steps are named by the functions. But I don't mind changing it back and adding comments instead.

lukash replied:
If you want to keep it split in multiple functions, fine with me. But do you also dislike to have a single function for that duplicate piece of code above the `for` loop and inside it?

kontura replied:
In the end I put it into a single bigger function with comments that also de-duplicates the code.

lukash commented on rgo/builders/copr.py:
```diff
@@ -53,39 +53,42 @@ def __init__(
             LOGGER.info("Using existing COPR project: %s/%s", self.project.ownername, self.project.name)
 
-            # add the build chroots to the project's chroots and extend its expiration time
+            self.all_project_chroots = set(list(self.project.chroot_repos.keys()) + self.cmdline_input_chroots)
```
This only sets `self.all_project_chroots` if the project exists (otherwise an exception is thrown above and you don't set it). You later rely on this being there, so I think it will fail if you call this on a non-existent project.

lukash commented on rgo/builders/copr.py:
```diff
@@ -161,19 +164,44 @@ def build(self, component, with_build_id=None, after_build_id=None):
+
+        # if we have atleast one default (not overriden) chroot do the default build
+        if (list(self.all_project_chroots - overriden_chroots)):
+            buildopts["chroots"] = list(self.all_project_chroots - overriden_chroots)
```
This just doesn't do what it's supposed to. The logic behind the chroots is this:
- If you don't specify `--chroots` on CLI, it builds for all chroots of the Copr project
- If you specify `--chroots` on CLI, it builds the chroots you've specified (and adds them to the Copr project if they weren't there before)
- If you don't specify `--chroots` and the Copr project doesn't exist yet, it errors out saying you need to specify the chroots for the new project

Here you override the previously set `buildopts["chroots"]` unconditionally, always building all project chroots.

kontura replied:
I see, I have misunderstood that, I though the chroots from command line are just added. WIll fix it.

lukash commented on rgo/builders/copr.py:
```diff
@@ -161,19 +164,44 @@ def build(self, component, with_build_id=None, after_build_id=None):
+
+        for override in component.overrides:
+            if override.roots:
+                buildopts["chroots"] = list(override.roots)
```
And here this unconditionally builds for all overridden chroots, which:
- Is not what is expected; I think what's expected is to build either `--chroots` specified on CLI or all Copr project chroots. Any overrides that are not on that list should just be ignored for this build (the overlay can be used in multiple places, you don't always want to build the chroots that have overrides)
- Will IMO break if you've got overrides for chroots that aren't configured in the Copr project.

kontura replied:
Should be hopefully working as expected now.

lukash commented on rgo/builders/copr.py:
```diff
@@ -161,19 +164,44 @@ def build(self, component, with_build_id=None, after_build_id=None):
+                    "after: %s " % after_build_id if after_build_id is not None else "",
+                    self._build_url(build.id)
+                ))
+                override.build_id = build.id
```
The code in this `for` loop is again very similar to the one called above for the non-overridden builds, it should go into a private method of this class. It's just a few calls, but the LOGGER line itself is quite non-trivial and shouldn't be duplicated.

lukash commented on rgo/builders/copr.py:
```diff
@@ -189,26 +217,20 @@ def wait_for_results(self):
+                    if not component.done:
+                        component.build_proxy = self.client.build_proxy.get(component.build_id)
+                        component.done, failed = utils.report_build_proxy_and_get_status(component.build_proxy,
+                                                                                         component.name)
```
This function can again be a private method of this class and you can put the proxy `get` call inside it as well.

kontura replied:
I will make it a private method of the class but I am not sure I like adding the `get` call in there. It would have to take the `component` and `override` as an argument and treat them the same even though they have a different type.

lukash commented on rgo/component.py:
```diff
@@ -95,94 +93,31 @@ def make_srpm(self, tmpdir):
+        final_spec_path = utils.prepare_spec_and_archive(workdir, original_spec_path, self.git,
+                                                         self.name, self.version_from, patches)
+        if patches == PatchesAction.keep:
+            utils.prepare_specifle_patches(workdir, final_spec_path, self.distgit)
```
Typo: specifle :slightly_smiling_face:

m-blaha commented on rgo/git.py:
```diff
@@ -360,3 +360,25 @@ def clone(self, dest):
+
+
+class Override(DistGit):
+    def __init__(self, src, freeze=None, branch=None, spec_path=None, roots=[],
```
Mutable default `roots=[]`.

kontura replied:
Fixed, thank you!

kontura committed:
Add make_srpms which uses make_srpm for default and override builds

kontura committed:
SRPMS: Rename variables, add comments, simplify patches handling

It also raises an exception if only a distgit is used since we cannot
handle it. There is no way to download tarball from distgit.

lukash commented on rgo/git.py:
```diff
@@ -360,3 +360,27 @@ def clone(self, dest):
+        """
+        super().__init__(src, freeze=freeze, branch=branch, spec_path=spec_path,
+                         patches=patches, patches_dir=patches_dir, type_=type_)
+        if roots is None:
```
Do you have a reason for calling this `roots`? It's called `chroots` in Copr and also in rest of the rpm-gitoverlay, let's keep it consistent?

kontura replied:
Done, I also changed it in the overlay it self to **chroots:**
```
...
    distgit-overrides:
      - chroots:
          - centos-stream-8-x86_64
        src: centos-distgit:libsolv.git
        branch: c8s
        spec-path: SPECS/libsolv.spec
        patches-dir: SOURCES/
```

lukash commented on rgo/component.py:
```diff
@@ -183,10 +181,40 @@ def make_srpm(self, tmpdir):
+        # Build distgit_overrides srpms
+        for distgit_override in self.distgit_overrides:
+            distgit_override.srpm = Component._make_srpm(tmpdir, distgit_override, self.git, self.name,
+                                                         self.version_from, distgit_override.patches)
```
The git, name and version_from arguments are the same for both calls, how about keeping the `_make_srpm()` method non-static and using the `self` attributes directly instaed of passing through arguments?

kontura replied:
ok, updated.

lukash commented on rgo/component.py:
```diff
@@ -98,32 +98,36 @@ def _make_srpm(tmpdir, distgit, source_code_git, pkg_name, version_from, patches
             spec_version = spec_version.decode("utf-8")
 
-        _spec_path = os.path.join(workdir, "{!s}.spec".format(spec_name))
+        _final_spec_path = os.path.join(workdir, "{!s}.spec".format(spec_name))
```
Since you're renaming... what's the point of having some variables prefixed with an underscore?

lukash commented on rgo/component.py:
```diff
@@ -98,32 +98,36 @@ def _make_srpm(tmpdir, distgit, source_code_git, pkg_name, version_from, patches
 
+        # PREPARE SPECFILE AND ARCHIVE
         if source_code_git:
+
```
Linebreak somewhat superfluous?

lukash commented on rgo/component.py:
```diff
@@ -135,40 +139,35 @@ def _make_srpm(tmpdir, distgit, source_code_git, pkg_name, version_from, patches
-            if src_type == rpm.RPMBUILD_ISPATCH:
-                if patches_action == PatchesAction.keep:
+            shutil.copy2(original_spec_path, _final_spec_path)
+            raise NotImplementedError("Getting a tarball from distgit is not implemented")
```
Is this a mistake? You're copying the spec and then raising an exception directly after that.

kontura replied:
I have commented it out with some description.

lukash commented on rgo/builders/copr.py:
```diff
@@ -53,39 +53,44 @@ def __init__(
             if chroot not in self.client.mock_chroot_proxy.get_list():
                 raise Exception("{!r} is not an active chroot".format(chroot))
-        self.chroots = chroots
+        self.cmdline_input_chroots = set(chroots)
```
Unless I'm missing something, I see this distinction of `cmdline_input_chroots` and `all_project_chroots` as an unnecessary complication. `self.chroots` simply represented the chroots you're building, which is the only thing you care about after figuring out this set, which is either what came from the CLI, or what is set on the project. No need to store both and then add another `if` when you're going to use the values in `build`.

The only difference seems to be `self.chroots` weren't set if they didn't come from CLI, you can set them to the project chroots for that case and the rest of the code can work with this one value.

kontura replied:
Yes you are right!
Thanks.

kontura committed:
Build overrides in Copr and report status of the builds

It also changes how the builds are tracked, we are keeping directly the
`build_proxy` response from Copr. This is basically a dictionary (which
Copr calls `Munch`) of the API response.

Since we want to store this for component's default build and also for
overrides it is a nice structure representing the build.

lukash commented:
Looking good now, thanks!
